gensim/models/doc2vec.py

```diff
@@ -847,7 +847,7 @@ def save_word2vec_format(self, fname, doctag_vec=False, word_vec=True, prefix='*
                     fout.write(utils.to_utf8("%s %s\n" % (total_vec, self.vector_size)))
                 # store as in input order
                 for i in range(len(self.docvecs)):
-                    doctag = prefix + str(self.docvecs.index_to_doctag(i))
+                    doctag = "%s%s" % (prefix, self.docvecs.index_to_doctag(i))
```

Comment:
We should pick one type (byte string or unicode) and stick with it.

Supporting both types at the same time like this (`doctag` will be silently upcast to unicode in this solution, if the argument is unicode) looks very brittle.

I'm not as familiar with this codebase as @gojomo is, but converting to one fixed type (bytestring or unicode) during input processing is preferable. At this point, during model save, the type ought to be fixed.

Reply:
I agree that the _hidden_ cast to unicode is not optimal. The problem is that the method [`index_to_doctag`](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/doc2vec.py#L332) can return a string as well as an int. The fix is similar to another place in the class [here](https://github.com/RaRe-Technologies/gensim/blob/develop/gensim/models/doc2vec.py#L395).
I am open for other suggestions though.

Reply:
Aha, so the expected types are either unicode or int. A `u"%s" % index_to_doctag` should work either way, right? No change needed (beside dropping that `str`). IIRC that int was an optimization hack (uses less memory).

My suggestion is to keep the tags explicitly as "unicode or int", or "bytestring or int". And then during string formatting, rely on the fact it's unicode-or-int (or byte-or-int), possibly even with an assert.

@gojomo What are the invariants here, how to do this cleanly? That unicode-or-str-or-int is really confusing, and apparently not encapsulated well (fixes in one place leave bugs in other places).

Reply:
IMO the type of `tags` should be as permissive as possible – conceivably even anything that works as a dict key would work everywhere in `Doc2Vec` *except* in this new bit of convenience for shoehorning doc-vecs into the (limited, somewhat-'legacy') word2vec.c format. For people who want to use this format, they should be sure to use tags that can write as a nice string-token – I'd think it'd be OK here if this code is tolerant of any reasonable such choice. (They may also need to make sure, themselves, that their tags don't include internal spaces if using the 'text' word2vec.c format.)

Reply:
@piskvorky I changed it to `u"%s%s" % ..` so that the cast is less hidden.
@gojomo Your last point is especially important, yeah.